PopPUNK/utils.py

```diff
@@ -497,3 +494,162 @@ def isolateNameToLabel(names):
+    # retain sketches of failed samples
+    if retain_failures:
+        hdf_fail = h5py.File(failed_db_name, 'w')
+    
+    # process data structures
+    read_grp = hdf_in['sketches']
+    if qc_filter == 'prune':
+        out_grp = hdf_out.create_group('sketches')
+    if retain_failures:
+        fail_grp = hdf_fail.create_group('sketches')
+    seq_length = {}
+    seq_ambiguous = {}
+    seq_excluded = {}
+    removed = []
+    retained = []
+    
+    # iterate through sketches
+    for dataset in read_grp:
+        # test thresholds
+        remove = False
+        seq_length[dataset] = hdf_in['sketches'][dataset].attrs['length']
+        seq_ambiguous[dataset] = hdf_in['sketches'][dataset].attrs['missing_bases']
+        # if no filtering to be undertaken, retain all sequences
+        if qc_filter == 'continue':
+            retained.append(dataset)
+
+    # calculate thresholds
+    with open(prefix + '/' + os.path.basename(prefix) + '_qcreport.txt', 'a+') as qc_file:
+    
+        # get mean length if not using user-supplied estimation
+        if estimated_length is None:
+            genome_lengths = np.fromiter(seq_length.values(), dtype = int)
+            mean_genome_length = np.mean(genome_lengths)
+        
+        # calculate length threshold unless user-supplied
+        if lower_length is None:
+            lower_length = mean_genome_length - length_sigma*np.std(genome_lengths)
+            upper_length = mean_genome_length + length_sigma*np.std(genome_lengths)
+
+        # iterate through and filter
+        example_assembly = None
+        failed_sample = False
+        for dataset in seq_length.keys():
+                    
+            # determine if sequence passes filters
+            remove = False
+            if seq_length[dataset] < lower_length:
+                remove = True
+                qc_file.write(dataset + '\tBelow lower length threshold\n')
+            elif seq_length[dataset] > upper_length:
+                qc_file.write(dataset + '\tAbove upper length threshold\n')
+            if upper_n is not None and seq_ambiguous[dataset] > upper_n:
+                remove = True
+                qc_file.write(dataset + '\tAmbiguous sequence too high\n')
+            elif seq_ambiguous[dataset] > prop_n * seq_length[dataset]:
+                remove = True
+                qc_file.write(dataset + '\tAmbiguous sequence too high\n')
+
+            # write to files
+            if remove:
+                sys.stderr.write(dataset + ' failed QC\n')
+                failed_sample = True
+                if retain_failures:
+                    fail_grp.copy(read_grp[dataset], dataset)
+            else:
+                if qc_filter == 'prune':
+                    out_grp.copy(read_grp[dataset], dataset)
+                    retained.append(dataset)
+                # retain an example assembly
+                if example_assembly is None:
+                    example_assembly = dataset
```

Comment:
Just pick the first one - stops there being issues with the conditionals here (it looks like it won't be defined if `remove = True`, and then l624 will error)

Reply:
Left over from avoiding picking an assembly that was filtered out.

Reply:
Done in e601f73.